File: Go/010-summation-of-primes.go
package main

import (
	"fmt"
)

//Main function
func main() {
	const top = 2000000
	total := uint64(0)
	// "sieve of eratosthenes"
	sieve := [top]bool{} // true when not prime
	for n := 2; n < top; n++ {
		if sieve[n] {
			// we marked this as a nonprime from a previous factor - not a prime
			continue
		}
		// we know it's a prime - add to our running total
		total += uint64(n)
		for nonprime := n * n; nonprime < top; nonprime += n {
			// mark all multiples of this prime as non primes (divisible by this prime)
			sieve[nonprime] = true
		}
	}

	fmt.Printf("the value is: %v\n", total)
}

File: Ruby/010-summation-of-primes.rb
# The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

# Find the sum of all the primes below two million.
def sum_primes_to(top)
  prime_sum = 0
  # "sieve of eratosthenes"
  sieve = Array.new(top, false) # true when not prime
  n = 2
  while n < top do
    # we marked this as a nonprime from a previous factor - not a prime
    if sieve[n]
      n += 1
      next
    end

    # we know it's a prime - add to our running total
    prime_sum += n
    nonprime = n * n
    
    # mark all multiples of this prime as non primes (divisible by this prime)
    while nonprime < top do
      sieve[nonprime] = true
      nonprime += n
    end
    n += 1
  end
  prime_sum
end

puts sum_primes_to(2000000)